WIP: Starknet Snap Rough Skeleton (#1509)
MetaMask/metamask-docs: wallet/how-to/use-non-evm-networks/starknet/manage-starknet-accounts.md

---
description: Manage Starknet accounts in MetaMask.
sidebar_position: 2
---

# Manage Starknet accounts

> ## Account creation

Reviewer: As we discussed during the sync, maybe move this section lower. Also, add a brief intro to this page.

Account creation in Starknet is handled by the wallet provider. As a dapp developer, you do not create accounts directly. Instead, you guide users to create an account with their preferred wallet provider.

:::note

Currently, multiple Starknet accounts are not supported in the Starknet Snap.

:::

## View account information

After a user connects, you can display the account details, such as the account address: 

> ```javascript

Reviewer: Please update this and all other code samples to use double quotes instead of single quotes.

  }
};
```

## Retrieve connected Starknet accounts

To retrieve and display connected Starknet accounts, use the `get-starknet` library in combination with React hooks:

```javascript
import { useStarknet, useConnectors } from '@starknet-react/core';
import { useState, useEffect } from 'react';

function AccountDisplay() {
  const { account } = useStarknet();
  const { available, connect, disconnect } = useConnectors();
  const [accountAddress, setAccountAddress] = useState<string | undefined>();

  useEffect(() => {
    setAccountAddress(account?.address);
  }, [account]);

  return (
    <div>
      {accountAddress ? (
        <p>Connected Account: {accountAddress}</p>
      ) : (
        <p>No account connected</p>
      )}
      {available.map((connector) => (
        <button key={connector.id()} onClick={() => connect(connector)}>
          Connect {connector.name()}
        </button>
      ))}
      {account && (
        <button onClick={() => disconnect()}>Disconnect</button>
      )}
    </div>
  );
}
```

This component displays the connected account address if available, and provides buttons to connect or disconnect accounts.

## Manage account transactions

> You can manage transactions with `get-starknet`:
> 
> ```javascript
> 

Reviewer: ````suggestion
You can manage transactions using `get-starknet` as follows:

```javascript
````

function AccountChangeHandler() {
  const [account, setAccount] = useState<string | null>(null);

  useEffect(() => {
    const starknet = getStarknet();

    const handleAccountsChanged = (accounts: string[]) => {
      console.log('Accounts changed:', accounts);
      setAccount(accounts[0] || null);
      // Update your app's state here
    };

    const handleDisconnect = () => {
      console.log('Disconnected from wallet');
      setAccount(null);
      // Handle disconnection (e.g., reset app state, show connect button)
    };

    if (starknet) {
      starknet.on('accountsChanged', handleAccountsChanged);
      starknet.on('networkChanged', handleDisconnect);

      // Initial account setup
      starknet.enable().then((accounts: string[]) => {
        setAccount(accounts[0] || null);
      });

      return () => {
        starknet.off('accountsChanged', handleAccountsChanged);
        starknet.off('networkChanged', handleDisconnect);
      };
    }
  }, []);

  return (
    <div>
      {account ? (
        <p>Connected Account: {account}</p>
      ) : (
        <p>No account connected</p>
      )}
    </div>
  );
}

export default AccountChangeHandler;
```

> Use this component at the top level of your app to handle account changes globally.

Reviewer: ```suggestion
Use this component at the top level of your dapp to handle account changes globally.
```

---
description: Manage Starknet accounts in MetaMask.
sidebar_position: 2
---

# Manage Starknet accounts

## View account information

After a user connects, you can display the account details, such as the account address: 

```javascript
const showAccountInfo = async () => {
  const account = await connectStarknetAccount();
  if (account) {
    document.getElementById('accountAddress').innerText = `Account Address: ${account}`;
  }
};
```

> ## Switch between accounts
> 
> With the `get-starknet` library, you can allow users to switch accounts by re-enabling the wallet. 
> MetaMask will prompt the user to select a different account if multiple accounts are available.
> 
> ```javascript
> const switchStarknetAccount = async () => {
>   try {
>     const starknet = getStarknet();
>     await starknet.enable();  // Prompts the user to select an account
>     const account = starknet.selectedAddress;
>     console.log('Switched to Starknet Account:', account);
>     
>     return account;
>   } catch (error) {
>     console.error('Error switching Starknet account:', error);
>   }
> };
> ```

Reviewer 2: We should remove as this is not yet implemented in Starknet Snap. 

```suggestion
```

    console.error('Error invoking contract:', error);
  }
};
```

# Comprehensive Guide to Managing Starknet Accounts

## 1. Retrieving and Displaying Connected Starknet Accounts

To retrieve and display connected Starknet accounts, you'll typically use the `get-starknet` library in combination with React hooks for state management.

```typescript
import { useStarknet, useConnectors } from '@starknet-react/core';
import { useState, useEffect } from 'react';

function AccountDisplay() {
  const { account } = useStarknet();
  const { available, connect, disconnect } = useConnectors();
  const [accountAddress, setAccountAddress] = useState<string | undefined>();

  useEffect(() => {
    setAccountAddress(account?.address);
  }, [account]);

  return (
    <div>
      {accountAddress ? (
        <p>Connected Account: {accountAddress}</p>
      ) : (
        <p>No account connected</p>
      )}
      {available.map((connector) => (
        <button key={connector.id()} onClick={() => connect(connector)}>
          Connect {connector.name()}
        </button>
      ))}
      {account && (
        <button onClick={() => disconnect()}>Disconnect</button>
      )}
    </div>
  );
}
```

> This component will display the connected account address if available, and provide buttons to connect or disconnect accounts.

Reviewer 2: Not sure this is needed as we already have something very similar at the beginning of the tutorial.

> ## 2. Switching Between Multiple Starknet Accounts

Reviewer 2: As mentioned earlier, things related to multiple starknet accounts are not yet supported in the Starknet Snap. We can remove this section for now.

  useEffect(() => {
    if (account?.address && !connectedAccounts.includes(account.address)) {
      setConnectedAccounts([...connectedAccounts, account.address]);
    }
  }, [account]);

  const switchAccount = async (connector: any) => {
    await disconnect();
    await connect(connector);
  };

  return (
    <div>
      <h2>Connected Accounts:</h2>
      <ul>
        {connectedAccounts.map((address) => (
          <li key={address}>{address}</li>
        ))}
      </ul>
      <h2>Available Connectors:</h2>
      {available.map((connector) => (
        <button key={connector.id()} onClick={() => switchAccount(connector)}>
          Switch to {connector.name()}
        </button>
      ))}
    </div>
  );
}
```

This component keeps track of all connected accounts and allows switching between available connectors.

## 3. Account Creation Process

Account creation in Starknet is typically handled by the wallet provider (e.g., ArgentX, Braavos). As a dApp developer, you don't create accounts directly. Instead, you guide users to create an account with their preferred wallet provider. Here's how you might implement a guide for users:

> ```typescript

Reviewer 2: Note that the `connect` method of `get-starknet` 

```
import { connect } from "get-starknet"

return <button onClick={() => connect()}>Connect wallet</button>
```

When clicking on the connect wallet button above the users will see the list of compatible wallet, will be able to click to install one or use one already installed. 

This interface can be modified if needed, more information is available on get-starknet GitHub

function AccountCreationGuide() {
  return (
    <div>
      <h2>Create a Starknet Account</h2>
      <ol>
        <li>Install a Starknet-compatible wallet (e.g., ArgentX, Braavos)</li>
        <li>Open the wallet and follow its account creation process</li>
        <li>Once created, return to this dApp and connect your new account</li>
      </ol>
      <p>After creating your account, you can connect it to this dApp.</p>
    </div>
  );
}
```

## Handling Account Changes and Disconnections

> To handle account changes and disconnections, you can use event listeners provided by get-start. Here's an example using `@starknet-react/core`:

Reviewer 2: The handling of disconnection is already present at the beginning of the tutorial in the handleDisconnect method. Maybe we could simply describe it here ?

  Instead, you can guide users to [create an account](connect-to-starknet.md) with MetaMask.
- Currently, the Starknet Snap doesn't support multiple Starknet accounts.

:::

## Prerequisites

[Connect to Starknet](connect-to-starknet.md) from your dapp.

## Display account information

After a user connects to their Starknet account in MetaMask, you can display the account details.
The following example displays the account address: 

<Tabs>
  <TabItem value="get-starknet" default>

  ```javascript
  const showAccountInfo = async () => {
    const account = await connectStarknetAccount();
    if (account) {
      document.getElementById("accountAddress").innerText = `Account Address: ${account}`;
    }
  };
  ```

  </TabItem>
  <TabItem value="wallet_invokeSnap">
  
  ```javascript
  const showAccountInfo = async () => {
    if (typeof window.ethereum !== "undefined" && window.ethereum.isMetaMask) {
      try {
        // Invoke the Starknet Snap to get account information.
        const response = await window.ethereum.request({
          method: "wallet_invokeSnap",
          params: {
            snapId: "npm:@starknet-snap/snap", 
            request: {

>               method: "starknet_getAccounts"

Reviewer: This method `starknet_getAccounts` is used in several examples but is not documented in the reference. Is there a spec for it?

---
description: Manage Starknet accounts in MetaMask.
sidebar_position: 2
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

> # Manage Starknet accounts

Reviewer: Since the examples on this page display information in the dapp, it would be nice to include screenshots. Can be raised for a future PR.

    const result = await starknet.account.execute({
      contractAddress: contractAddress,
      entrypoint: entrypoint,
      calldata: calldata
    });

    console.log("Transaction result: ", result);
  } catch (error) {
    console.error("Error invoking contract:", error);
  }
};
```

  </TabItem>
  <TabItem value="wallet_invokeSnap">

  ```javascript
  const invokeStarknetContract = async () => {
  if (typeof getEip6963Provider !== "undefined" && getEip6963Provider.isMetaMask) {
    try {
      const contractAddress = "0xYourContractAddress";  // Replace with your contract address.
      const entrypoint = "function_name";  // The function you want to call.
      const calldata = [/* your function arguments */];

      const result = await getEip6963Provider.request({
        method: "wallet_invokeSnap",
        params: {
          snapId: "npm:@starknet-snap/snap",
          request: {
            method: "starkNet_executeTxn",
            params: {
              contractAddress: contractAddress,
              contractFuncName: entrypoint
              contractCallData: calldata.join(","),
              senderAddress: "0xb60e8dd61c5d32be8058bb8eb970870f07233155", 

>               maxFee: "1000000000000000", //

Reviewer: Should these be updated to match the listed `executeTxn` params?
